File: Ruby/depth-first-search.rb
# Define Node class

class Node
  attr_accessor :data, :left, :right

  def initialize(data)
    @data = data
    @left = nil
    @right = nil
  end

  def to_s
    print "Node with data value of #{self.data}" + "\n"
  end
end

# Define depth_first_search method

def depth_first_search(node, value)
  return node if node.data == value
  left = depth_first_search(node.left, value) if node.left
  right = depth_first_search(node.right, value) if node.right
  (left || right) || "Value not found"
end

# Build tree

head_node = Node.new(5)
head_node.left = Node.new(6)
head_node.right = Node.new(8)
head_node.left.left = Node.new(78)
head_node.left.right = Node.new(87)
head_node.right.left = Node.new(4)
head_node.right.right = Node.new(9)

# Run method

print depth_first_search(head_node, 77)

File: Javascript/depth-first-search.js
// Define node constructor function

function node(data) {
  this.data = data;
  this.left = null;
  this.right = null;
}

// Define depthFirstSearch function

function depthFirstSearch(node, value) {
  if (node.data === value) {
    return node;
  }
  if (node.left) {
    if (depthFirstSearch(node.left, value)) {
      return depthFirstSearch(node.left, value);
    } else {
      return depthFirstSearch(node.right, value);
    }
  }
}

// Build tree

var headNode = new node(5);
headNode.left = new node(4);
headNode.right = new node(2);
headNode.left.left = new node(25);
headNode.left.right = new node(87);
headNode.right.left = new node(65);
headNode.right.right = new node(3);

// Run the function

console.log(depthFirstSearch(headNode, 65));